### ISSUE TITLE
Follow up to LXE translation PR

### ISSUE THREAD

**Issue by sbcgua** [abapGit/abapGit#4470]
to do after #4415 

1) Major. the langs are saved as a table of sy-langu. This has 2 issues

This
![image](https://user-images.githubusercontent.com/15635498/106360743-eeb7e880-6322-11eb-8262-4ebdce9fe885.png)

And the fact is that langu is a customizing, while my initial code used ISO codes there, which are far more stable.
So I propose to revert it to ISO codes (and maybe to string form, but not sure, it has + and -)

2) Double pull required. 
I also confirm this issue. Currently needs double pull (pull + force pull) to deserialize properly.
I think that the proper way would be to run LXE deserialization as a post-step after activation. But that will probably need some refactoring ...

3) Minor. Currently dotabapgit class expose 2 ways to interact with translation list. We should leave one (table-like).

4) `zcl_abapgit_lxe_texts=>convert_lang_string_to_table` and `convert_table_to_lang_string` raise on missing language. But languages may be missing in a system. This should not block the deserialization of code and supported langs. I'd suggest soften the check as it was initially.

5) Duplication of texts. This is a long term stuff. The master lang LXE data is not used anyhow. It is there because it was in the data structure. So uses extra space. Probably makes sense to look how to optimize it. Theoretically, a specific object may know the structure of it's own LXEs and extract the master lang data to fill the gaps during actual object deserialization, so that notmal object texts are not needed. OR, in opposite, fill LXE data from regular object texts before LXE deserialization (probably better to keep compatibility with previous XML structure).

**Comment by mbtools** [comment]
1. `<MASTER_LANGUAGE>` is sy-langu (one letter). "Inside SAP" things work with the one-letter format. If for some reason (like LXE) something needs a different format, then convert at the time of need (see 3. as well)

I would give the `i18n_languages` a proper structure to replace the "ugly"  `<SYST--2DLANGU>`.

2. Yes. Remove `de/serialize_lxe_texts()` from the serializer classes and make it universal i.e. post activation.

3. Removing the `get/set_i18n_lang` methods goes in hand with improving the settings dialog. Typing in the language codes is asking for trouble. It should be replaced by a picklist. For example, make the input field read-only and select languages in a popup. Or we put the language settings on a separate tab to have the i18n languages in a multi-select list control.

4. But how can any language be missing in T002 (unless someone maliciously has deleted it)?

5. I would keep the original language part of the LXE node in XML because it's necessary to perform a translation. It just makes it that much simpler to universally extract the texts and put them back with translations, later.  [We're talking about a few Kbytes. I wouldn't worry about size of repo XMLs at all. Just adding an image to the readme would be more than all texts combined].

**Comment by sbcgua** [comment]
1. MASTER_LANGUAGE happened historically, it does not mean that it is right because it is like this. When we speak about translations, ISO codes are much more clear and universal. And readable too. Do you know what is language `8` for example or `A` or `a` (mind the case - they differ ;)? Also Master language is guaranteed in the system. While translation are not. I think it is really a mistake to store 1-char SAP keys in a repo (which is relatively abstract). And will cause issues.
2. Yeap, kind of. But where the data will live all this time ? I think there will be some challenges. Anyway, a task to think on.
3. "...select languages in a popup ..." - mmm, for seems like a complication. On the other hand guarantees some overview of installed/missing languages. Why not.
4. There is just 45 entries in T002 in NW752-dev. For sure fewer than ISO lang codes. In our dev system it's 41. So using 4 langs from NW752 will dump in our system

(5) Duplication of texts. Did you try serializing more than 2 langs ? I bet the english part is repeated for each lang. Does it add some contrast ? Anyway I think this question even deserves a separate ticket. It might be a not an easy part to solve anyway. Maybe even does not worth the effort for real cases. I deal with just 2-3 additional languages, I think I can tolerate some extra XML size :) But technically - not ideal.

**Comment by mbtools** [comment]
Devil's advocate: What language is ISO "4F"?  ;-)

The XMLs in the repo are serialized ABAP data. Nothing is in an external format. I don't think we want users to edit the XML and therefore have provided a UI to maintain the settings. So why make an exception for this one field?

If a language doesn't exist in the target system, you should get an error (not a dump). It could be a warning. I think the deserialize should then be adapted because storing for example French texts in a German system where French is not installed is questionable (not sure if the LXE functions would complain). 

OK, I understand the duplicate original text issue if you have several translations. There's room for improvement, indeed.

**Comment by sbcgua** [comment]
> What language is ISO "4F"?

I don't know. But I can go here - https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes - and check. Seems like there is not such code - they are letter only :) And this will be true for all the systems, irrespective to their customizing.

P.S. T002X seems to have dfltlang field by the way. And we need to check how sap1_to_sap2 converts stuff to be sure.

So to summarize:
Pros of ISO codes:
- readable
- does not depend on case
- depend on strict ISO standard not customizing

Pros of LANGU
- hmmm, close to abap internals ?

Let me think a bit what is the better choice ... 

> If a language doesn't exist in the target system, you should get an error (not a dump). It could be a warning. I think the deserialize should then be adapted because storing for example French texts in a German system where French is not installed is questionable (not sure if the LXE functions would complain).

Exactly ! Warning would be nice, I agree. But it should not block deserialization. So that check should be moved elsewhere.

**Comment by sbcgua** [comment]
Does AG have a warning mechnism to be called from deserialization by the way ?

**Comment by mbtools** [comment]
Three options:
- Upfront messages in `zcl_abapgit_object=>deserialize_check` (ie. `warning...` methods)
- Upfront messages from `zcl_abapgit_file_status=>run_checks' 
- Messages to include in the log when using `deserialize` which users would see only if they view the log afterward

**Comment by sbcgua** [comment]
updated the issue description

**Comment by mbtools** [comment]
6. the texts between quotes are not visible
7. I would do a warning when entering the repo view
8. Since LXE is still experimental, the idea was to keep the old i18n methods as the default and only use LXE if translation languages are maintained in settings. once LXE is central and working for all objects, this switch in PROG and FUGR can be removed.
9. I saw that too. In some cases, it doesn't matter (like perf test). Agree. it would be better to have consistent calls. `ZCL_ABAPGIT_SERIALIZE->CONSTRUCTOR` should just import `IS_I18N_PARAMS type zif_abapgit_definitions=>ty_i18n_params` and pass it down to the serializer class (and get rid of the various language parameters). I would have done it already but there are too many changes in the queue.
10. I couldn't reproduce this. New issue?

**Comment by sbcgua** [comment]
> the texts between quotes are not visible

oops, fixed. I meant `*`

**Comment by mbtools** [comment]
6. I would remove the `*` logic completely. It's not intuitive. Better to add a value help for translation language which shows all installed languages (except main language). Select all by default with an option to deselect some.

**Comment by JustACasual** [comment]
>     1. Rethink `*` usage. Problem: we can set `*` in original system, it makes sense - will save all installed langs. But what about receiving system ? The real list of langs is missing - so there is no way to process them properly (e.g. find intersection between installed and repo langs). Proposal is to convert `*` into real list of langs on settings save. So that .abapgit contain real languages only and not `*`. @JustACasual What do you think ?

Sorry, not much time to spare at the moment :( But I will try to react on mentions anyway
Could you elaborate about the exact problem?
Here is what I would expect for the * to do:
1st system: 
- * is set, DE, EN, FR is installed, all the texts exist
- XML contains an entry for each of these languages

2nd system: 
- * is set, DE, EN, ES is installed
- import XML
- will deserialize DE and EN,
- will not deserialize FR as not installed (just catch error?)
- will probably show a diff always for the FR texts
- if ES text will be maintained, it will be serialized. FR will not be serialized, as not installed

when pulling from 1st system again, will probably show diffs for FR and for ES
-> as stated earlier on in the discussion, those diffs are fine for me :)

If people are messing around with different language sets and just set "give me everything" (aka *) and then texts are translated into these very languages, I think this should be diffs. Because it will really create differences in the objects (as in 'diffs').

**Comment by sbcgua** [comment]
@JustACasual yes, all you wrote was correct. My idea was just to avoid saving `*` to .abapgit. Instead make it a UI feature. So user wants to add all installed languages. He does not know the exact list. He enters `*` (or presses a button as Marc suggested) and the list is autocreated. But the `*` is not saved anywhere.

So if you don't mind we'll be going that direction.

**Comment by JustACasual** [comment]
Ah ok sorry, then I got you wrong!
Yes sounds good!

**Comment by mbtools** [comment]
New twist in this story. I somehow missed what @DerGuteWolf built:
https://blogs.sap.com/2021/02/09/abap-report-using-sap-translation-hub-for-i18n-of-texts-of-abap-artefacts/
https://github.com/DerGuteWolf/ABAP_TRANSLATION_HUB

I haven't tried the whole round trip, but it makes me rethink the complete i18n approach we tried to tackle here. 

Instead of adding the LXE data to each object XML, it should be handled separately - kind of like we do with TABU and the /data/ folder. Serialize LXE data into something that the translation hub understands - an XLIFF `.xlf` file (it's an XML format) - into an /i18n/ folder in the repo. One file per source/target language combination makes sense.

ABAP_TRANSLATION_HUB is using `cl_lxe_textext_ex/import` and the logic and XML formats are actually very similar to what we have in `zcl_abapgit_lxe_texts`. 

Anyway, just something to consider.
